### PR TITLE
Enable some mypy options

### LINKED ISSUE #177: Enable some mypy options
### What is the expected enhancement?

Some mypy options look useful. How about enabling following options?
https://mypy.readthedocs.io/en/stable/command_line.html

- [--no-implicit-optional](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-optional)
  > This flag causes mypy to stop treating arguments with a None default value as having an implicit Optional type.
- [--warn-redundant-casts](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-warn-redundant-casts)
  > This flag will make mypy report an error whenever your code uses an unnecessary cast that can safely be removed.
- [--warn-unused-ignores](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-warn-unused-ignores)
  > This flag will make mypy report an error whenever your code uses a `# type: ignore` comment on a line that is not actually generating an error message.

### PR DESCRIPTION
**Pull request by manoelmarques**
### Summary
Fixes #177

### CODE REVIEW

**Comment by t-imamichi** [comment]
Thank you for setting up github actions. I found the three errors only for Python 3.6 due to the new config. The first one is introduced by #175. We need it for Python 3.7 and 3.8.
```
qiskit_optimization/algorithms/admm_optimizer.py:685: error: unused "type: ignore" comment
qiskit_optimization/applications/tsp.py:158: error: unused "type: ignore" comment
qiskit_optimization/applications/tsp.py:174: error: unused "type: ignore" comment
```

It might be difficult to use the new mypy options to keep the consistent mypy results for all Python 3.6-3.8.

**Comment by manoelmarques** [comment]
@t-imamichi I changed the CI to create a configuration file on the fly for python 3.6 without the new options. However I agree it is confusing if someone tries runs mypy on python 3.6. I tried it in qiskit-nature with the new options and had 23 errors.

I believe we should hold on until we drop python 3.6 support.

You can feel free to close this PR and the issue.

**Comment by t-imamichi** [comment]
Thank you for trying a new configuration. Let's leave this on hold.

**Comment by t-imamichi** [comment]
Shall we enable this PR once Python 3.6 is deprecated?

**Comment by manoelmarques** [comment]
Yes, once the PR to eliminate python 3.6 from this repository is merged (after the 3.0 release) this PR can be merged too.

**Comment by t-imamichi** [comment]
These new mypy options detected two cases.
```
qiskit_optimization/converters/inequality_to_equality.py:270: error: Redundant cast to "float"
test/applications/test_bin_packing.py:100: error: unused "type: ignore" comment
```

File: qiskit_optimization/algorithms/admm_optimizer.py [hunk 1]
```diff
@@ -682,7 +682,7 @@ def _create_step3_problem(self) -> QuadraticProgram:
         quadratic_y = self._params.beta / 2 * np.eye(binary_size) + self._state.rho / 2 * np.eye(
             binary_size
         )
-        op3.objective.quadratic = quadratic_y  # type: ignore[assignment]
```
Comment (t-imamichi) [inline qiskit_optimization/algorithms/admm_optimizer.py:685]: Thanks! I forgot about this.

**Review by t-imamichi — APPROVED** [review]

**Pull request merged**